src/mmtk.rs

```diff
@@ -220,6 +220,90 @@ impl<VM: VMBinding> MMTK<VM> {
         }
     }
 
+    /// Initialize the GC worker threads that are required for doing garbage collections.
+    /// This is a mandatory call for a VM during its boot process once its thread system
+    /// is ready.  This function should be called exactly once, and should be called before
+    /// allocating objects in the MMTk heap.
```

Comment:
That makes sense, but I think the current API is a bit wound up w.r.t. `GlobalState::initialized` and `VM::VMCollection::is_collection_enabled()`.  If I read the code correctly, the current code does assume `GlobalState::initialized` to be true when allocating (if it is possible to trigger GC).  But we may do another refactoring to tidy up the interacction between `initialized` and `is_collection_enabled`.

`Space::acquire()` contains the following code:

```rs
        let should_poll = VM::VMActivePlan::is_mutator(tls) && VM::VMCollection::is_collection_enabled();
        let allow_gc = should_poll && self.common().global_state.is_initialized();

        if should_poll && self.get_gc_trigger().poll(false, Some(self.as_space())) {
            assert!(allow_gc, "GC is not allowed here: collection is not initialized (did you call initialize_collection()?).");
```

If `should_poll` (determined by `is_collection_enabled()`) is true, and `poll()` indicates that collection is required, it will assert `allow_gc` is true.  That will fail if `initialized` is false.  If we allow allocation before calling `initialize_collection()`, we will also require `is_collection_enabled()` to always return false before calling `initialize_collection()`, too.  That requirement has not yet been reflected by our API docs, but I think it complicates things anyway.

`memory_manager::gc_poll()` also asserts `is_initialized()` after `VM::VMCollection::is_collection_enabled()` returns `true`.  It is arguable that `gc_poll` is for triggering GC, not allocation, and that we can require the VM not to call `gc_poll()` before calling `initialize_collection()`, but it is not that obvious from the API.

There are also some places involving stress GC that call `is_initialized()` (but strangely they do not `is_collection_enabled()`).  Examples include `Allocator::alloc_slow_inline` and `GCTrigger::should_do_stress_gc_inner`.

So I think for the current code, it is the safest to require `initialize_collection()` before allocation.  But I'll raise another issue for that.

Reply:
I don't think we want to require `initialize_collection()` before allocation. If our code requires that, it sounds like a bug to me. The current behavior is that the bindings can allocate before `initialize_collection()`, but when the heap is full and the binding has not yet called `initialize_collection()`, it will fail. What we may want is probably that a binding can allocate as much as they want without triggering a GC before calling `initialize_collection()`, which is the same as Java MMTk.

Reply:
OK.  For now, I can remove "and should be called before allocating objects in the MMTk heap" from the comment.

Reply:
I just encountered the assertion failure in Julia. I created an issue here https://github.com/mmtk/mmtk-core/issues/1096.